### PR TITLE
Fixes #8158 - Multiple entries with same key for aliases in elasticsearch connector

### LINKED ISSUE #8158: elasticsearch query bug
when I query elasticsearch use trino, there is  a bug
my query is :
`SELECT table_schema FROM columns WHERE elasticsearch.information_schema.table_schema = 'default'`
then i meet the bug: Multiple entries with same key
In my elasticsearch, there are many indexes alias one name, eg:
index |  alias_name
aa-000000 -->aa
aa-000001 -->aa
i think  there is some bug in io.trino.plugin.elasticsearch.ElasticsearchMetadata.listTable method

### PR DESCRIPTION
**Pull request by sumannewton**
Fixes #8158 

The issue, RCA & fix is explained in this comment - https://github.com/trinodb/trino/issues/8158#issuecomment-860181923

Thanks.

### CODE REVIEW

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java [hunk 1]
```diff
@@ -347,10 +347,10 @@ else if (type instanceof ObjectType) {
                 .map(index -> new SchemaTableName(this.schemaName, index))
                 .forEach(result::add);
 
-        client.getAliases().entrySet().stream()
-                .filter(entry -> indexes.contains(entry.getKey()))
-                .flatMap(entry -> entry.getValue().stream()
-                        .map(alias -> new SchemaTableName(this.schemaName, alias)))
+        client.getAliases().entrySet()
```
Comment (martint) [inline plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java:350]: `getAliases` doesn't need to change, and all you need is to restructure this to dedup the names before constructing the list. You can do it like this:

```java
        Set<SchemaTableName> result = new HashSet<>();
        
        indexes.stream()
                ...
                .forEach(result::add);

        client.getAliases().entrySet().stream()
                ...
                .forEach(result::add);

        return ImmutableList.copyOf(result);
```

**Review by martint — CHANGES_REQUESTED** [review]
Please add a test in BaseElasticsearchConnectorTest

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java (hunk 1 above)
Comment (sumannewton) [reply plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java:350]: Currently `getAliases()` computes and returns Map(Index, List(Alias)). So whoever fetches aliases in the future also, will have to dedup every time. But with this PR, we compute and return Map(Alias, List(Indices)), which makes more sense. No need to dedup. No need to iterate through the entire Map to identify all the indices of an alias.

Is there any specific significance why we return Map(Index, List(Alias)) currently?

Comment (bitsondatadev) [reply plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java:350]: I think @martint wants to keep these calls aligned with the structure that Elasticsearch returns to make the client simple to understand for others. `getAliases()` and other calls in the client are in essence dumb wrappers that just maps the rest call response to in-mem Java structure. From there it's up to us to map it however we like. If this is going to add a lot of extra work, then we can cache the another structure in the client or somewhere else if necessary. It doesn't look like a lot to me though.

**Review by bitsondatadev — COMMENTED** [review]
Let's go with what @martint suggests and leave `getAliases()` as is and update to dedup in the `ElasticsearchMetadata` method.

Also could you add a test here please. Thanks again for all this work to improve the es connector!

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java (hunk 1 above)
Comment (sumannewton) [reply plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java]: Got it. Only dedup is handled now.

**Comment by sumannewton** [comment]
@bitsondatadev @martint Addressed review comment. Requesting your review again. Thanks.

**Review by bitsondatadev — COMMENTED** [review]
@sumannewton could you please add a test that verifies this behavior exists to avoid unintended changes or side effects from elasticsearch versions in the future?

**Comment by sumannewton** [comment]
> @sumannewton could you please add a test that verifies this behavior exists to avoid unintended changes or side effects from elasticsearch versions in the future?

@bitsondatadev Shouldn't this test be added as part of BaseConnectorTest. Because this issue can happen in any connector. And currently, we have no test in BaseConnectorTest that validates this query.

**Comment by sumannewton** [comment]
> @sumannewton could you please add a test that verifies this behavior exists to avoid unintended changes or side effects from elasticsearch versions in the future?

@bitsondatadev Added test.

**Review by bitsondatadev — COMMENTED** [review]
One quick update to test

File: plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java [hunk 2]
```diff
@@ -1017,6 +1017,18 @@ public void testAlias()
...
     }
 
+    @Test
+    public void testSelectInformationSchemaForMultiIndexAlias()
+            throws IOException
+    {
+        addAlias("nation", "multi_alias");
+        addAlias("region", "multi_alias");
+
+        // No duplicate entries should be found in information_schema.tables or information_schema.columns.
```
[the remark this answers is not in the document]
Comment (sumannewton) [reply plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1027]: `testSelectInformationSchemaColumns` test actually does `select ... from information_schema.columns where table_schema = .....` This ensures that no duplicates are available. To be safe I added `testSelectInformationSchemaTables` test too.

Also when these 2 tests ran before, aliases weren't present hence they passed. So I added a test for alias and retriggered the same internally to unit test it.

[the remark this answers is not in the document]
Comment (sumannewton) [reply plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1027]: I tried adding this test without the fix, it failed with the same error reported in #8158. Post fix, test passed.

[the remark this answers is not in the document]
Comment (bitsondatadev) [reply plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:1027]: Ah, okay well in that case this test works! I'll go ahead an approve it!

**Review by bitsondatadev — APPROVED** [review]
Looks good!

**Comment by martint** [comment]
Looks good. One minor comment: can you change the commit message to say "Fix failure when indexes share same alias" and add a short summary to the commit message describing what was the problem?

**Comment by sumannewton** [comment]
> Looks good. One minor comment: can you change the commit message to say "Fix failure when indexes share same alias" and add a short summary to the commit message describing what was the problem?

@martint All comments addressed.

**Comment by bitsondatadev** [comment]
Thanks @sumannewton 🎉

**Comment by bitsondatadev** [comment]
I'm getting the following Assertion errors when running the elasticsearch tests for `testSelectInformationSchemaForMultiIndexAlias` locally.


```
java.lang.AssertionError: For query: 
 SELECT table_name, column_name FROM information_schema.columns WHERE table_catalog = 'elasticsearch' AND table_schema = 'tpch' AND table_name LIKE '%orders%'
not equal
Actual rows (up to 100 of 9 extra rows shown, 18 rows in total):
    [orders_alias, clerk]
    [orders_alias, comment]
    [orders_alias, custkey]
    [orders_alias, orderdate]
    [orders_alias, orderkey]
    [orders_alias, orderpriority]
    [orders_alias, orderstatus]
    [orders_alias, shippriority]
    [orders_alias, totalprice]
Expected rows (up to 100 of 0 missing rows shown, 9 rows in total):
	at org.testng.Assert.fail(Assert.java:94)
	at io.trino.testing.QueryAssertions.assertEqualsIgnoreOrder(QueryAssertions.java:237)
	at io.trino.testing.QueryAssertions.assertQuery(QueryAssertions.java:203)
	at io.trino.testing.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at io.trino.testing.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:167)
	at io.trino.testing.BaseConnectorTest.testSelectInformationSchemaColumns(BaseConnectorTest.java:664)
	at io.trino.plugin.elasticsearch.BaseElasticsearchConnectorTest.testSelectInformationSchemaForMultiIndexAlias(BaseElasticsearchConnectorTest.java:1192)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:110)
```

**Comment by bitsondatadev** [comment]
Just ran the test singled out and it passed. The test failure above occurred with all tests running. Going to run all tests locally again and see if I can replicate.

**Comment by bitsondatadev** [comment]
Seems to be failing when running other tests together. Strangely this didn't occur on github test runs so I have asked Suman to see if he can replicate locally to see if it's an issue on my end.

**Comment by phd3** [comment]
@bitsondatadev thanks for reporting: https://github.com/trinodb/trino/pull/8379#issue-677302781

**Comment by bitsondatadev** [comment]
No problem, I think @sumannewton is working on a fix. May want to sync with him.

**Pull request merged**